llvm/include/llvm/IR/RepoTicket.h

```diff
@@ -40,6 +40,10 @@ static constexpr ticketmd::DigestType NullDigest{
 using GOVec = SmallVector<const GlobalObject *, 1>;
 /// Map GO to a unique number in the function call graph.
 using GOStateMap = llvm::DenseMap<const GlobalObject *, unsigned>;
+/// Memoize GO's hash if one of the following conditions is satisfied:
+/// 1. GO is not in a cycle/loop.
+/// 2. GO contains an edge back to itself.
```

Comment:
An edge back to itself sounds like a loop so this reads (to me) like:
(GO is not in a loop) OR (GO is in a loop)

which equates to true.

Reply:
Paul (from email):

Yes it is. The algorithm has a special case for “self loops” so that they don’t prevent a vertex (and everything connecting to it) from being cached. In other words: in a graph like A -> B -> B, we can cache the results for both A and B even though there’s a loop. In contrast, if the loop were A -> B -> A, then neither can be cached.

Reply:
I think that makes it clearer.

Reply:
FTR, I've expanded the description of the original algorithm [here](https://github.com/paulhuggett/digraph-hash/blob/master/README.md) so that it talks about this detail. I thought that this information was already there but was mistaken.

Reply:
Can we avoid the associative array by adding a digest (and "bool memoized" flag) to GlobalObject?

I'm not sure whether that would be faster or not.

Reply:
I did compare the different implementations (which include the one your mentioned here) before this code review. A new flag named Cached is added into the TicketNode metadata. The existing digest in the TicketNode is used to keep the cached digest. The TicketNode metadata is only used by the repo target (not by other targets). I found this method got worse performance compared the associate array.
Since the digest and cached flag is stored in the TicketNode, I used the following code to get it:
```
GO->getMetadata(LLVMContext::MD_repo_ticket))
```
I found the GO’s metadata is stored in the densemap as below:
```
DenseMap<const GlobalObject *, MDGlobalAttachmentMap> GlobalObjectMetadata;
```
`MDGlobalAttachmentMap` is a multimap-like storage for metadata attachments for globals. When using the `getMetadata` function to get the GO’s TicektNode. There are two main steps: 1) get GO’s attachments (a SmallVector of all metadata) from GlobalObjectMetadata; 2) get GO’s TicketNode metadata from MDGlobalAttachmentMap. We can see this method would be slower.

Reply:
I have updated the comments.